appengine/oauth2/src/main/java/com/example/appengine/Oauth2Filter.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.appengine;
+
+import static com.google.appengine.api.utils.SystemProperty.environment;
+
+import com.google.appengine.api.oauth.OAuthRequestException;
+import com.google.appengine.api.oauth.OAuthService;
+import com.google.appengine.api.oauth.OAuthServiceFactory;
+import com.google.appengine.api.oauth.OAuthServiceFailureException;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.utils.SystemProperty;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ *  Filter to verify that request has a "Authorization: Bearer xxxx" header,
+ *  and check if xxxx is authorized to use this app.
+ *
+ *  Note - this is to demonstrate the OAuth2 APIs, as it is possible to lockdown some
+ *  of your app's URL's using cloud console by adding service accounts to the project.
+ */
+public class Oauth2Filter implements Filter {
+
+  private ServletContext context;
+
+  @Override
+  public void init(final FilterConfig config) throws ServletException {
+    this.context = config.getServletContext();
+  }
+
+  // [START oauth2]
+  @Override
+  public void doFilter(final ServletRequest servletReq, final ServletResponse servletResp,
+                             final FilterChain chain) throws IOException, ServletException {
+    final String scope = "https://www.googleapis.com/auth/userinfo.email";
+    Set<String> allowedClients = new HashSet<>();
```

Comment:
Could we use an ImmutableSet here and add the clients in the builder? That is, does `allowedClients` need to be mutable? If not, we could even make this a `private static final` constant.